mlir/lib/Dialect/Rock/Transforms/Regularize.cpp

```diff
@@ -147,187 +146,452 @@ struct RegularizeGenericRewritePattern
+  if (!maybeBufferReaders)
+    return allocOp.emitError(
+        "couldn't find an operation that reads this buffer");
+  SmallVector<OpOperand *> bufferWriters = std::move(*maybeBufferWriters);
+  SmallVector<OpOperand *> bufferReaders = std::move(*maybeBufferReaders);
+  if (bufferWriters.size() != 1) {
+    allocOp->getParentOp()->dump();
+    return allocOp.emitError(
+               "expected one operation to write this buffer, but there are ")
+           << bufferWriters.size();
+  }
+
+  OpOperand *writer = bufferWriters[0];
+  bool isInputFusion = state.inputFusionWrites.contains(writer);
+  if (isInputFusion)
+    return pushTransformsToInputFusionReaders(allocOp, bufferDeps, writer,
+                                              state, rw);
+  if (!state.outputFusionWrites.contains(writer)) {
+    return allocOp.emitOpError("couldn't find the writer of a buffer in the "
+                               "set of analyzed fusable writes. See ")
+           << bufferWriters.front()->get();
+  }
+
+  OpOperand *transformingReader = nullptr;
+  for (OpOperand *reader : bufferReaders) {
+    Value readArg = reader->get();
+    if (!readArg.getDefiningOp<rock::TransformOp>()) {
+      continue;
     }
-    return lres;
+    if (transformingReader && transformingReader != reader)
+      return isolateMultipleTransformingReaders(
+          allocOp, writer->getOwner(), bufferDeps, bufferReaders, state, rw);
+    transformingReader = reader;
   }
-};
+  // It's the simple case, do nothing
+  if (!transformingReader)
+    return success();
+
+  if (transformingReader && bufferReaders.size() > 1) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Adding copies to partially intercept a transform stack\n");
+    return isolateMultipleTransformingReaders(
+        allocOp, writer->getOwner(), bufferDeps, bufferReaders, state, rw);
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "Processing output fusion read via "
+                          << transformingReader->get() << " (argument "
+                          << transformingReader->getOperandNumber() << " of "
+                          << *transformingReader->getOwner() << ") via "
+                          << allocOp << "\n");
+
+  return regularizeTransformStack(allocOp, transformingReader, bufferDeps,
+                                  state, rw);
+}
+
+LogicalResult findFusionRoots(func::FuncOp kernel,
+                              const BufferDependencyAnalysis &bufferDeps,
+                              TransformPushState &state) {
+  bool foundFusionRoot = false;
+  kernel.walk([&](Operation *op) {
+    llvm::TypeSwitch<Operation *>(op)
+        .Case(
+            [&](memref::AllocOp allocOp) { state.worklist.push_back(allocOp); })
+        .Case([&](GridwiseGemmOp gemmOp) {
+          foundFusionRoot = true;
+          collectOutputFusionWrites(&gemmOp.getCMutable(), bufferDeps, state);
+          collectInputFusionWrites(&gemmOp.getAMutable(), bufferDeps, state);
```

Comment:
See all these are buffers; not readers or writers. Therefore, collectXFusionWrites API is a bit confusing...

Reply:
For example, in a GEMM, the A and B arguments are readers - they have MemRead effects - and C is a writer

Reply:
A,B, and C are all buffers or view of them. They are not readers or writers.

To me reader is something like : 
```mlir
%buffer1 = memref.alloc()
%buffer2 = memref.alloc()
threadwise_read_into %buffer1 -> %buffer2
threadwise_write_all %buffer2 -> %arg
``

So `threadwise_read_into` is a reader of `%buffer1`.
The first OpOperand of  `threadwise_read_into` is ` %buffer1` that is actually a buffer or a view of it.
The second OpOperand of  `threadwise_read_into` is ` %buffer2` that is actually a buffer or a view of it.
None of the OpOperands, I would not classify as readers and writers.
Hence my confusion.